File: shiny/driver/windriver/windraw.c
// Copyright 2015 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include "_cgo_export.h"
#include "windriver.h"

static HBITMAP mkbitmap(HDC dc, RECT *r, VOID **ppvBits) {
	BITMAPINFO bi;
	LONG dx, dy;
	HBITMAP bitmap;

	dx = r->right - r->left;
	dy = r->bottom - r->top;

	ZeroMemory(&bi, sizeof (BITMAPINFO));
	bi.bmiHeader.biSize = sizeof (BITMAPINFOHEADER);
	bi.bmiHeader.biWidth = (LONG) dx;
	bi.bmiHeader.biHeight = -((LONG) dy); // negative height to force top-down drawing
	bi.bmiHeader.biPlanes = 1;
	bi.bmiHeader.biBitCount = 32;
	bi.bmiHeader.biCompression = BI_RGB;
	bi.bmiHeader.biSizeImage = (DWORD) (dx * dy * 4);

	bitmap = CreateDIBSection(dc, &bi, DIB_RGB_COLORS, ppvBits, 0, 0);
	if (bitmap == NULL) {
		// TODO(andlabs)
	}
	return bitmap;
}

static void blend(HDC dc, HBITMAP bitmap, RECT *dr, LONG sdx, LONG sdy) {
	HDC compatibleDC;
	HBITMAP prevBitmap;
	BLENDFUNCTION blendfunc;

	compatibleDC = CreateCompatibleDC(dc);
	if (compatibleDC == NULL) {
		// TODO(andlabs)
	}
	prevBitmap = SelectObject(compatibleDC, bitmap);
	if (prevBitmap == NULL) {
		// TODO(andlabs)
	}

	ZeroMemory(&blendfunc, sizeof (BLENDFUNCTION));
	blendfunc.BlendOp = AC_SRC_OVER;
	blendfunc.BlendFlags = 0;
	blendfunc.SourceConstantAlpha = 255;  // only use per-pixel alphas
	blendfunc.AlphaFormat = AC_SRC_ALPHA; // premultiplied
	if (AlphaBlend(dc, dr->left, dr->top,
		dr->right - dr->left, dr->bottom - dr->top,
		compatibleDC, 0, 0, sdx, sdy,
		blendfunc) == FALSE) {
		// TODO
	}

	// TODO(andlabs): error check these?
	SelectObject(compatibleDC, prevBitmap);
	DeleteDC(compatibleDC);
}

// TODO(andlabs): Upload

void fillSrc(PVOID dc, RECT *r, COLORREF color) {
	HBRUSH brush;

	// COLORREF is 0x00BBGGRR; color is 0xAARRGGBB
	color = RGB((color >> 16) & 0xFF,
		(color >> 8) & 0xFF,
		color & 0xFF);
	brush = CreateSolidBrush(color);
	if (brush == NULL) {
		// TODO
	}
	if (FillRect(dc, r, brush) == 0) {
		// TODO
	}
	// TODO(andlabs) check errors?
	DeleteObject(brush);
}

void fillOver(PVOID dc, RECT *r, COLORREF color) {
	HBITMAP bitmap;
	VOID *ppvBits;
	RECT oneByOne;

	// AlphaBlend will stretch the input image (using StretchBlt's
	// COLORONCOLOR mode) to fill the output rectangle. Testing
	// this shows that the result appears to be the same as if we had
	// used a MxN bitmap instead.
	oneByOne.left = 0;
	oneByOne.top = 0;
	oneByOne.right = 1;
	oneByOne.bottom = 1;
	bitmap = mkbitmap(dc, &oneByOne, &ppvBits);
	*((uint32_t *) ppvBits) = color;
	blend(dc, bitmap, r, 1, 1);
	// TODO(andlabs): check errors?
	DeleteObject(bitmap);
}

// TODO(andlabs): Draw

File: shiny/driver/windriver/windraw.go
// Copyright 2015 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// +build windows

package windriver

import (
	"syscall"
	"unsafe"

	"golang.org/x/exp/shiny/driver/internal/win32"
)

func mkbitmap(dx, dy int32) (syscall.Handle, *byte, error) {
	bi := _BITMAPINFO{
		Header: _BITMAPINFOHEADER{
			Size:        uint32(unsafe.Sizeof(_BITMAPINFOHEADER{})),
			Width:       dx,
			Height:      -dy, // negative height to force top-down drawing
			Planes:      1,
			BitCount:    32,
			Compression: _BI_RGB,
			SizeImage:   uint32(dx * dy * 4),
		},
	}

	var ppvBits *byte
	bitmap, err := _CreateDIBSection(0, &bi, _DIB_RGB_COLORS, &ppvBits, 0, 0)
	if err != nil {
		return 0, nil, err
	}
	return bitmap, ppvBits, nil
}

func blend(dc win32.HDC, bitmap syscall.Handle, dr *_RECT, sdx int32, sdy int32) (err error) {
	compatibleDC, err := _CreateCompatibleDC(dc)
	if err != nil {
		return err
	}
	defer func() {
		err2 := _DeleteDC(compatibleDC)
		if err == nil {
			err = err2
		}
	}()
	prevBitmap, err := _SelectObject(compatibleDC, bitmap)
	if err != nil {
		return err
	}

	blendfunc := _BLENDFUNCTION{
		BlendOp:             _AC_SRC_OVER,
		BlendFlags:          0,
		SourceConstantAlpha: 255,           // only use per-pixel alphas
		AlphaFormat:         _AC_SRC_ALPHA, // premultiplied
	}
	err = _AlphaBlend(dc, dr.Left, dr.Top,
		dr.Right-dr.Left, dr.Bottom-dr.Top,
		compatibleDC, 0, 0, sdx, sdy,
		blendfunc.ToUintptr())
	if err != nil {
		return err
	}

	_, err = _SelectObject(compatibleDC, prevBitmap)
	return err
}

func blit(dc win32.HDC, dp _POINT, bitmap syscall.Handle, sr *_RECT) (err error) {
	compatibleDC, err := _CreateCompatibleDC(dc)
	if err != nil {
		return err
	}
	defer func() {
		err2 := _DeleteDC(compatibleDC)
		if err == nil {
			err = err2
		}
	}()
	prevBitmap, err := _SelectObject(compatibleDC, bitmap)
	if err != nil {
		return err
	}

	dx, dy := sr.Right-sr.Left, sr.Bottom-sr.Top
	_, err = _BitBlt(dc, dp.X, dp.Y, dx, dy, compatibleDC, sr.Left, sr.Top, _SRCCOPY)
	if err != nil {
		return err
	}
	_, err = _SelectObject(compatibleDC, prevBitmap)
	return err
}

func fillSrc(hwnd win32.HWND, uMsg uint32, wParam, lParam uintptr) {
	dc, err := win32.GetDC(hwnd)
	if err != nil {
		panic(err) // TODO handle error?
	}
	defer win32.ReleaseDC(hwnd, dc)
	r := (*_RECT)(unsafe.Pointer(lParam))
	color := _COLORREF(wParam)

	// COLORREF is 0x00BBGGRR; color is 0xAARRGGBB
	color = _RGB(byte((color >> 16)), byte((color >> 8)), byte(color))
	brush, err := _CreateSolidBrush(color)
	if err != nil {
		panic(err) // TODO handle error
	}
	defer _DeleteObject(brush)
	err = _FillRect(dc, r, brush)
	if err != nil {
		panic(err) // TODO handle error
	}
}

func fillOver(hwnd win32.HWND, uMsg uint32, wParam, lParam uintptr) {
	dc, err := win32.GetDC(hwnd)
	if err != nil {
		panic(err) // TODO handle error
	}
	defer win32.ReleaseDC(hwnd, dc)
	r := (*_RECT)(unsafe.Pointer(lParam))
	color := _COLORREF(wParam)

	// AlphaBlend will stretch the input image (using StretchBlt's
	// COLORONCOLOR mode) to fill the output rectangle. Testing
	// this shows that the result appears to be the same as if we had
	// used a MxN bitmap instead.
	bitmap, bitvalues, err := mkbitmap(1, 1)
	if err != nil {
		panic(err) // TODO handle error
	}
	defer _DeleteObject(bitmap) // TODO handle error?
	*(*_COLORREF)(unsafe.Pointer(bitvalues)) = color
	if err = blend(dc, bitmap, r, 1, 1); err != nil {
		panic(err) // TODO handle error
	}
}

var (
	msgFillSrc  = win32.AddWindowMsg(fillSrc)
	msgFillOver = win32.AddWindowMsg(fillOver)
	msgUpload   = win32.AddWindowMsg(handleUpload)
)

// TODO(andlabs): Draw
